### DISCUSSION TITLE
Add a more comfortable installation method as AppImage does.

### DISCUSSION THREAD

**Discussion by jefer94** [microsoft/winget-cli#4506] [General]
### Description of the new feature / enhancement

The installation process is so complex for some users with things like, where to install the files, folders, path, user agreements of 7.000.000 pages to read, 17 toolbars to be added to Internet Explorer (Currently fixed), 3 new antivirus for you, 9 new apps that could enjoy in the installation process and that you won't use.

Sometimes it does not include the required dependencies or any way to claim to the developer if it won't open even if you paid for that software, and I don't know, but I see the installation process some older today, I mean in Windows, I have this problem and in Linux is so complicated for the most users, and in Mac the steps literally are 1. download it 2. move it to the apps folder and 3. open it and enjoy.

It should also add more metadata to the installed programs to support new features like remembering all the apps that had been activated and should be installed automatically as Android does.

### Proposed technical implementation details

I think that in the next major release of Windows, it must include an installation process that every elder could understand easily. I think that this should be this way 1. download it, 2. double click in it 3. say just yes or no, and 4. to have installed there only a new program, not 3, 9 or 65 programs, just 1.

The program should be enveloped and have inside it a simple folder structure, and it should use a secondary folder for dependencies like `dependencies/Abc-1.0.2` copying how NPM or any package manager works it should prevent dependency issues using a manifest.

**Comment by Masamune3210** [comment]
A. That already exists. It's called MSIX.
B. This isn't really the place for OS Feature requests. I somewhat doubt that the winget team has the pull needed to do major feature requests simply out of the blue, let alone the time.

**Reply by denelon** [reply]
Yes, logically the MSIX packaging format is very similar to the second part of the proposed technical implementation details. You can actually download one, rename it to .zip, expand the archive, and you can observe an XML based manifest along with the files necessary to install the package. In some cases, there are also references to dependencies (frameworks) which are automatically distributed via the Microsoft Store.

In some cases, the applications can be side loaded using the App Installer or manually downloaded and installed by the user, and in those cases, the dependencies aren't automatically sourced, so some manual intervention is required.

One of the common terms used to refer to the MSIX applications is "packaged". Notionally, there are also "unpackaged" applications which refers to the earlier formats for installers, and currently there is no mechanism for a packaged application to take a dependency on an unpackaged dependency with an explicit declaration in the MSIX manifest. Over the fullness of time this may be addressed, or the scenario may remain unchanged. There are many arguments for both approaches. These are largely philosophical rather than technically limited. The current mechanism WinGet uses to deal with this, and support users is via the WinGet package manifest such that WinGet can install the unpackaged dependency as a required part of the "parent" package that could be packaged.

**Comment by jefer94** [comment]
Is it? I mean, I haven't seen any installer in Windows that works as I said, I suppose that you're telling me about the format that Windows Store uses, all installers that I have been watching are in .exe or .msi, remember that I told, download the file, just say yes or no, and enjoy, if it exists, I suggest you drop the others options and only accept them in compatibility mode for games.

If you are referring to that it happens through Microsoft Store, there is so much software not available and not uploaded to it, so, "it must include an installation process that every elder could understand easily" is not covered yet, also "to support new features like remembering all the apps that had been activated and should be installed automatically as Android does" directly couldn't be implemented because I recently installed windows 11 and this just not happened, I mean, when I bought a new phone, I don't have to install all my apps manually, it just works out the box, it installed my apps for me, it knew me, it knew what I installed.

Winget is a Microsoft product, if they don't want to have this here, they should move this issue opening it to another side, I mean, it's not so hard to do.

I think that they could cover this issue because Winget download the installer and runs it graphically, this is not accurate, I installed like 3 programs with it and this happened, doing this you should fix the installation process through Winget.

Microsoft is not very demanding with its developers and that is wrong, this affects the UX, I think that Microsoft must find a balance between not being as destructive as Apple (OpenGL, Vulkan, Epic Game case) and not being as permissive as Microsoft.

**Reply by denelon** [reply]
WinGet offers three installation modes "silent", "silent with progress", and "interactive". The default mechanism is "silent with progress" which allows the application to display UI, but it must not require user interaction. If a silent install is desired the "--silent" argument can be passed to suppress output, but this is still constrained by the installer itself, and some installers do not support a silent mechanism.

The Microsoft Store supports legacy UWP apps (these happen to be "packaged" MSIX applications) and Win32 apps (these are actually installed by WinGet through integration).

There is a "backup" and "restore" mechanism for users who have been running a Windows 10 or Windows 11 based PC using the Microsoft Account (MSA). When a new Windows 11 installation occurs for an MSA user, the option to restore from a previous device is given.  It's not as complete as what is described for Android, but we're working towards enabling that kind of an experience. The Windows ecosystem for software is a bit more complex in some cases than Linux based operating systems. Over the fullness of time, I would expect a similar delightful experience to happen, but there are many reasons due to complexity and priorities that this isn't already an option.

**Comment by stephengillie** [comment]
> AppImage is a format for packaging and running Linux applications without installation or root rights.

This is already covered by the `portable` package type.

**Comment by Masamune3210** [comment]
Projects already exist to add a GUI to control Winget. You can't just refuse people the ability to install non-"new fancy tech" programs and expect it to work out well for you. Look how much people loved Windows 8 RT, S-Mode, and whatever they decided to name it this week. 

Hint: People despise it.

Making a program for Windows has, is, and always should be a free action. Demanding to devs that they do it the new way or its the highway is just going to cause people to jump in their cars....

**Comment by jefer94** [comment]
Ok, what do you offer to fix the mentioned issues? and why don't use the Microsoft Store as the frontend of Winget?

**Comment by Masamune3210** [comment]
Because Winget supports any executable on the internet that can be downloaded from a direct link. Store does not. 

As for your mentioned issues:
Winget runs the installer with the silent switch set for the reported installer type. What the installer does with that information is fully up to it. If it does pop anything up, its only to show progress unless the installer is ill-behaved like EA's crappy installer, for instance.
Winget does not handle automatic installation after a reinstall, no, at least not by itself it doesn't. It does support exporting the currently installed and identified programs though and importing them to install them all in an automated fashion. If you wanted, you could just export with a scheduler task and then import whenever you need to reinstall.

**Comment by jefer94** [comment]
Download an installer from an unknown source could de dangerous, the other package managers have the concept of the repository, I suppose that this won't added to the Store because the Store is a product that takes a percentage of the revenue, but the most logical approach is that it shows the software with a warning or something like that, sincerely to have a different Frontend from the logical option isn't right.

I read the comment about portable apps, like Team Viewer and I've to say something about it, even if you got a portable app, it doesn't integrate with the start menu, app list, and the finder, or at least it is that I think.

And this point is about freedom, ok, that's ok, but which freedom? the user should choose what to consume, it's their machine, and some people and corporations think that it is ok to have a way to confuse, bend, and manipulate the users, and then they got weird tools and the result is the opposite to their will, I would suppose that the will and the freedom of the user matter, right?, corporations think that it is ok to have an agreement with 7.000.000 to read and that it is completely natural, a thing that everybody hate and it ends like the HumancentiPad episode, the freedom is limited when we affect third parties, I could extend this talking about the freedom of writing viruses but this topic is so easy that does not require explanations.

I've been reading about I'm talking about fancy tools or something like that, actually, I'm talking about something as simple as user experience, their will, and freedom, and also about having some experiences available in Android on Windows, that's it.

**Comment by stephengillie** [comment]
> Download an installer from an unknown source could de dangerous

The installer has to come from the developer, and not a 3rd party source. Additionally, a SHA-256 hash of the package is taken and stored, as a tamper-evident protection. Packages coming from sites that allow rehosting have to be reviewed by a human. Please read our policies before complaining, so your complaints can be accurate and addressable.

**Comment by jefer94** [comment]
That is the standard but even in that case, it could be dangerous (see the last attempt to attack Linux with a dependency that downloaded a malicious code), I don't talk about that because we understand the topic.

I see that the confusion is related to the end of that sentence, "repository", if I'm talking about repositories, I could be talking about adding repositories like APT, and I said about "unknown source" so, the repository could contain installers made for a stranger who signs its installer with SHA-256, and if this isn't the interpretations that you did, why would I say with this? "I suppose that this won't added to the Store because the Store is a product that takes a percentage of the revenue", if I'm talking about the possibility of loss of revenue it shows clearly that you can't get revenue for that repository for be a repository form a third party, so, this repository should contain infected installers.

I don't understand the confusion, I put the period at the end of the text where appeared the sentence you quoted.

**Reply by denelon** [reply]
When a user adds a new WinGet source, the default behavior in Windows to warn the user about the package coming from the Internet (or any untrusted zone), a warning is displayed to the user. Adding a new source to WinGet requires administrative privileges due to the potential risk inherent in trusting any third party. The exception to this is when an IT policy adds the source and specifies the source should be trusted. The Mark of the Web (MotW) mechanism is used to handle this, and the typical way this is managed is by "unblocking" files that come from an untrusted zone.

There is a difference between a "signed" application, and the SHA256 WinGet uses to determine if the installer matches the one downloaded by the URL referenced in the manifest. These are independent layers of security that users or IT organizations can leverage, or override based on their needs and decisions.

Publishers may choose to sell their products through the Microsoft Store, and in those cases, the publisher enters an agreement with Microsoft which may allow Microsoft to share in the revenue. Publishers may choose to offer their products through the Microsoft Store for "free" in which case, there is no revenue to "share". It is also possible for publishers to offer their products through the Microsoft Store and capture revenue for themselves without Microsoft being involved in the transaction. The publisher has choices.

**Comment by jefer94** [comment]
I think that at this point all the ideas are clear, and the next step must be the decision of Microsoft.

**Comment by denelon** [comment]
@Jefer94,

There is a lot to unpack here.

I don't believe all of the different topics are suitable for a "single" issue targeted towards WinGet. I'm going to convert this Issue to a discussion to help split some of the concerns out so I can coordinate communications with other teams at Microsoft who can address some of the discrete concerns.

**Comment by denelon** [comment]
I'm assuming you are referring to [AppImage](https://appimage.org/). I haven't used that tool, but the goal written in terms of "user stories" ` (as a <role> I want <behavior> so that <value>) ` seems to align nicely with similar intent to WinGet.

WinGet was originally designed and intended to help developers manage software on Windows. Since then, it's expanded beyond something exclusive to developers. There are what I would call power users who also enjoy the benefits and simplicity of a command-line interface. There are also several other IT professionals who are leveraging WinGet through custom integrations, as well as other Microsoft products integrated with WinGet like Microsoft Intune, Microsoft Dev Box, etc.

The default "sources" in WinGet are the Microsoft Store (with some limitations on the catalog content based on their policies) via the "msstore" source, and the community repository where essentially anyone can request a package to be added to the "winget" source. WinGet also offers an ability for users or organizations to add other sources.

There are GUI based applications like Microsoft Dev Home, and WinGetUI (soon to be renamed to UniGet) offering an alternative to the command line.

WinGet was designed to "meet developers where they are". This means most application types can be installed like a "portable" executable, and .exe-based installer, an MSI based installer, or an MSIX based installer. WinGet isn't doing any re-packaging to modify what the developer provides, and users can verify the provenance of those packages using ` winget show <package> ` which exposes several key pieces of metadata from the manifest.
